certbot/compat.py

```diff
@@ -0,0 +1,104 @@
+Instead, we apply a logic, and then fallback to another logic if first logic
+is not supported at runtime.
+
+Then logic chains are abstracted into single functions to be exposed to certbot.
+"""
+import os
+import select
+import sys
+import errno
+import ctypes
+
+from certbot import errors
+
+try:
+    # Linux specific
+    import fcntl # pylint: disable=import-error
+except ImportError:
+    # Windows specific
+    import msvcrt # pylint: disable=import-error
+
+def raise_for_non_administrative_windows_rights():
+    """
+    On Windows, raise if current shell does not have the administrative rights.
+    Do nothing on Linux.
+    """
+    if hasattr(ctypes, 'windll'):
+        windll = getattr(ctypes, 'windll')
+        if windll.shell32.IsUserAnAdmin() == 0:
+            raise ValueError('Error, certbot must be run on a shell with administrative rights.')
+
+def os_geteuid():
+    """Get current user uid"""
+    try:
+        # Linux specific
+        return os.geteuid()
+    except AttributeError:
+        # Windows specific
+        return '0'
+
+def readline_with_timeout(timeout, prompt):
+    """Read user input to return the first line entered, or raise after specified timeout"""
+    try:
+        # Linux specific
+        #
+        # Call to select can only be done like this on UNIX
+        rlist, _, _ = select.select([sys.stdin], [], [], timeout)
+        if not rlist:
+            raise errors.Error(
+                "Timed out waiting for answer to prompt '{0}'".format(prompt))
+        return rlist[0].readline()
+    except OSError:
+        # Windows specific
+        #
+        # No way with select to make a timeout to the user input on Windows,
+        # as select only supports socket in this case.
+        # So no timeout on Windows for now.
+        return sys.stdin.readline()
+
+def lock_file(fd):
+    """Lock the file linked to the specified file descriptor"""
+    if 'fcntl' in sys.modules:
+        # Linux specific
+        fcntl.lockf(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+    else:
+        # Windows specific
+        msvcrt.locking(fd, msvcrt.LK_NBLCK, 1)
+
+def release_locked_file(fd, path):
+    """Remove, close, and release a lock file specified by its file descriptor and its path."""
+    # Linux specific
+    #
+    # It is important the lock file is removed before it's released,
+    # otherwise:
+    #
+    # process A: open lock file
+    # process B: release lock file
+    # process A: lock file
+    # process A: check device and inode
+    # process B: delete file
+    # process C: open and lock a different file at the same path
+    try:
+        os.remove(path)
+    except OSError as err:
+        if err.errno in (errno.EACCES, errno.EPERM):
+            # Windows specific
+            #
+            # On Windows we cannot remove a file before closing its file descriptor.
+            # So we close first, and be exposed to the concurrency problem
+            # described in Linux section.
```

Comment:
First, do you think of the following code, to avoid misunderstanding on your remark :
```
try:
  os.remove(path)
except OSError as err:
  if err.errno in (errno.EACCESS, errno.EPERM)
    # Windows specific
    # We will not be able to remove a file before closing it.
    # To avoid race conditions described for Linux, we will not delete the lockfile,
    # just close it to be reused on the next Certbot call.
    pass
  else:
    raise
finally:
  os.close(fd)
```

But if we do not remove the lockfile, even if we release the lock, it will let some dirt in the directories processed by Certbot. Some Python libraries around there seems to have solved the issue both for Linux and Windows. I have already some work done for the next PR about the locking mechanism.

I think we can talk more about that in the relevant PR, but here is an excerpt: while Linux handles that correctly, the current locking process in Certbot leads to resource leak on Windows when Certbot finishes in a unexpectedly state. On AppVeyor. the VM is simply stale forever because the file and its lock are never released after failed tests that invoke the lock mechanism. I would like to construct a plateform-independent context manager for lock file, similar to thread locks or file descriptos and their `with` statement approach, fall-backed by the existing `atexit` in Certbot, to ensure that no lock or file descriptor ever leak for any exit state.

So I planned some modifications on the third PR for that to improve further the code. In the mean time, if what you thought is what I wrote below, I will integrate it in my PR.

Reply:
Integrated

Reply:
Yeah the code looks great now! Sorry I wasn't able to respond sooner.

That's awesome to hear you have ideas for how to improve locking on Windows. Feel free to keep implementing it, but if you want to write up what you're thinking to make sure we like the approach you have in mind before you spend more time on it, please do so and I'll try and respond quickly.

Reply:
In fact, I have already two PR ready after this one, one for the tests, one for the lock ^^ But it does not matter, I spent way more time to repair tests on windows than reimplementing a lock file mechanism.

So if you want a "preview" of the next PRs, you can check adferrand/certbot#5 (for the tests) and adferrand/certbot#6 (for the lock). They are constructed to display their own diffs, not the accumulated diff from certbot master.

Reply:
Oh wow. Nice work.

If they're already ready, I'll wait to review them like normal, but feel free to let me know in the future if you have something specific you'd like our opinion on early.